js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,385 @@
+    interest.refreshNonce();
+
+    this.fetchSegment(interest);
+
+    this.nextSegmentToRequest += 1;
+    this.increaseNumberOfSegmentsOnFly();
+  }
+};
+
+PipelineFixed.prototype.cancelPendingInterestsAboveFinalBlockId = function ()
+{
+  var len = this.dataFetchersContainer.length;
+  var i = this.finalBlockId;
+
+  for (var i = this.finalBlockId; i < len; i++) {
+    if (this.dataFetchersContainer[i] !== null) {
+      this.dataFetchersContainer[i].cancelPendingInterest();
+    }
+  }
+};
+
+PipelineFixed.prototype.onData = function(originalInterest, data)
+{
+  if (this.validatorKeyChain != null) {
+    try {
+      var thisPipeline = this;
+      this.validatorKeyChain.verifyData
+        (data,
+         function(localData) {
+           thisPipeline.onVerified(localData, originalInterest);
+         },
+         this.onValidationFailed.bind(this));
+    } catch (ex) {
+      console.log("Error in KeyChain.verifyData: " + ex);
+    }
+  }
+  else {
+    if (!this.verifySegment(data)) {
+      this.reportError(PipelineFixed.ErrorCode.SEGMENT_VERFIFICATION_FAILED,
+                       "Segment verification failed");
+    }
+
+    this.onVerified(data, originalInterest);
+  }
+};
+
+PipelineFixed.prototype.onVerified = function(data, originalInterest)
+{
+  var currentSegment = 0;
+
+  if (!PipelineFixed.endsWithSegmentNumber(data.getName())) {
+    // We don't expect a name without a segment number.  Treat it as a bad packet.
+    this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+                     "Got an unexpected packet without a segment number: " +
+                      data.getName().toUri());
+  }
+  else {
+    try {
+      currentSegment = data.getName().get(-1).toSegment();
+    }
+    catch (ex) {
+      this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+                       "Error decoding the name segment number " +
+                       data.getName().get(-1).toEscapedString() + ": " + ex);
+      return;
+    }
+  }
+
+  // if segment number of the very first Data packet does not start
+  // with zero, we have to start asking from segment zero
+  if (this.firstSegmentIsReceived === false) {
+    this.firstSegmentIsReceived = true;
+    if (currentSegment === 0) {
+      this.nextSegmentToRequest += 1;
+    }
+  }
+
+  // set finalBlockId only once
+  if ((this.finalBlockId === Number.MAX_SAFE_INTEGER) && (data.getMetaInfo().getFinalBlockId().getValue().size() > 0)) {
+    try {
+      this.finalBlockId = (data.getMetaInfo().getFinalBlockId().toSegment());
+      this.cancelPendingInterestsAboveFinalBlockId();
+    }
+    catch (ex) {
+      this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+           "Error decoding the FinalBlockId segment number " +
+            data.getMetaInfo().getFinalBlockId().toEscapedString() +
+            ": " + ex);
+      return;
+    }
+  }
+
+  if (this.satisfiedSegments[currentSegment] === true) {
```

Comment:
`satisfiedSegments` is redundant. If `contentParts` contains the segment number as a key, it means the segment is satisfied.